Add actions to allow modification of tasks

MartinBasti opened the pull request:
New subparser mod-task added with actions to modify task

eskultety commented:
> New subparser mod-task added with actions to modify task

nitpick: Maybe `modify` could be used as the sub-command's name due to a slightly easier spelling (since there likely won't be any bash completion)? After all, is there an expectation to modify anything else apart from tasks?

MartinBasti commented:
> > New subparser mod-task added with actions to modify task
> 
> nitpick: Maybe `modify` could be used as the sub-command's name due to a slightly easier spelling (since there likely won't be any bash completion)? After all, is there an expectation to modify anything else apart from tasks?

IMO in future we may need to update also pipeline parts, we may need to update params there too

MartinBasti commented:
> > > > New subparser mod-task added with actions to modify task
> > > 
> > > 
> > > nitpick: Maybe `modify` could be used as the sub-command's name due to a slightly easier spelling (since there likely won't be any bash completion)? After all, is there an expectation to modify anything else apart from tasks?
> > 
> > 
> > IMO in future we may need to update also pipeline parts, we may need to update params there too
> 
> `modify` could probably be a good choice and use subcommand or required optional argument to specify the resource to be modified, for example:
> 
> ```
> pmt modify --resource task add-param ...
> pmt modify --resource workspace ...
> pmt modify --resource pipeline add-param ...  # 
> pmt modify --resource result add ...
> ```
> 
> In this way, there will be less subcommands for modifying individual parts of a pipeline. Full spell "modify" should be better than "mod" anyway.

we already have add-task subcommad, so it follows the patter, also without the subparsers, we cannot provide different options, params will have differetn options than run after and workspaces

tkdchen commented on src/pipeline_migration/actions/mod_task.py:
```diff
@@ -0,0 +1,155 @@
+                            new_val = "\n".join(indented_lines)
+                            # TODO: this works by random magic only, it only assumes that  output will be exactly 2 lines
+                            # lc exists only in node, there is no lc in dict keys, thus we don't know exact structure
+                            # more heuristics is needed
```
I'm not sure I get your point here, especially for the `exact structure`. Maybe `YAMLStyle.detect` and `BlockSequenceIndentation.is_consistent are helpful.

For this case of adding a task parameter, I think we don't need to care about the `lc` of dict keys. By taking this example:

```yaml
tasks:
- name: clone
  params:
  - name: git-url
  - name: revision
- name: build
```

 as mentioned above `tasks[0]["params"]` is a `CommentedSeq` object, for adding parameter `revision`, its indentation is the `tasks[0]["params"].lc.col`.

tkdchen reviewed (commented):
Focusing on `mod_task.py` firstly. I'll continue reviewing the `fileoperations.py`

MartinBasti commented:
> > > > > New subparser mod-task added with actions to modify task
> > > > 
> > > > 
> > > > nitpick: Maybe `modify` could be used as the sub-command's name due to a slightly easier spelling (since there likely won't be any bash completion)? After all, is there an expectation to modify anything else apart from tasks?
> > > 
> > > 
> > > IMO in future we may need to update also pipeline parts, we may need to update params there too
> > 
> > 
> > `modify` could probably be a good choice and use subcommand or required optional argument to specify the resource to be modified, for example:
> > ```
> > pmt modify --resource task add-param ...
> > pmt modify --resource workspace ...
> > pmt modify --resource pipeline add-param ...  # 
> > pmt modify --resource result add ...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In this way, there will be less subcommands for modifying individual parts of a pipeline. Full spell "modify" should be better than "mod" anyway.
> 
> we already have add-task subcommad, so it follows the patter, also without the subparsers, we cannot provide different options, params will have differetn options than run after and workspaces
> 
> EDIT: I meant --resource cannot be used, it must be subparser
> 
> ```
> > pmt modify task add-param ...
> > pmt modify workspace ...
> > pmt modify pipeline add-param ...  # 
> > pmt modify result add ...
> ```

Updated PR, this argparrser is currently implemented
```
    pmt modify task TASK-NAME add-param PARAM-NAME PARAM-VALUE
    pmt modify task TASK-NAME remove-param PARAM-NAME
```

I reworked lot of code, created  generic module to help with yaml manipulations that can be reused in other resources

Please do the high level review only, if you like the concept, code is pure garbage will continue with improvents, but it somehow works (TODO: linters, datatypes, docs, logging)

The most critical missing part is handling yaml flow/inline representation intead of blocks.

MartinBasti commented on src/pipeline_migration/yamleditor.py:
```diff
@@ -0,0 +1,376 @@
+        path_stack.append((current_data, None))  # terminal node
+        return path_stack
+
+    def insert(self, path: YAMLPath, data: Any):
+        """Insert data into mapping or sequence, parent node must be specified as path"""
+        path_stack = self._get_path_stack(path)
+        last_node, _ = path_stack[-1]
```
Can you elaborate more what do you mean?

tkdchen commented on src/pipeline_migration/actions/modify/task.py:
```diff
@@ -67,7 +83,18 @@ def register_cli(subparser) -> None:
+        "-t",
+        "--type",
+        dest="param_type",
+        help=f"parameter type (Default: {ParamType.string})",
```
```suggestion
        help=f"parameter type (Default: %(default)s)",
```

MartinBasti replied:
fixed

MartinBasti commented on src/pipeline_migration/actions/modify/task.py:
Reworded commit message as follows:


pmt modify: support for yaml flow

With YAML flow, only limited support is implemented. If YAML flow style
is detected then it will be modified at the first non-yaml style parent.
All affected path will be regenerated, thus we cannot guarantee that
original style will be kept.

For example, adding item `{"item": "test"}` into flow style list:
```
---
start:
  flow: [{item: one, ...}, {item: two, ...}]
```

will result into:
```
---
start:
  flow:
  - {item: one, ...}
  - {item: two, ...}
  - item: test
```
As `start` was the first block yaml item, the whole `start` object will
be regenerated, flow mapping will be generated as block, new item will
be added as a block. The rest of items in `flow` will be kept as flow 
style as we didn't change their content.


YAML flow style is supported as the best effort.

MartinBasti commented on src/pipeline_migration/yamleditor.py:
```diff
@@ -0,0 +1,376 @@
+YAMLPath: TypeAlias = Sequence[Union[int, str]]
+
+
+class EditYAMLEntry:
```
updated

MartinBasti commented on src/pipeline_migration/yamleditor.py:
```diff
@@ -0,0 +1,376 @@
+            return isinstance(parent, dict)
+        return False
+
+    def _get_next_entry_line(self, path_stack):
```
added

MartinBasti commented on src/pipeline_migration/actions/modify/task.py:
```diff
@@ -0,0 +1,331 @@
+
+    pmt modify task push-dockerfile \\
+        -f /path/to/repo1/.tekton/pr.yaml -f /path/to/repo2/.tekton/push.yaml \\
+        add-param new-param new-value
```
fixed

MartinBasti commented on src/pipeline_migration/actions/modify/task.py:
```diff
@@ -0,0 +1,311 @@
+        if self._remove_param(tasks, ["spec", "pipelineSpec", "tasks"], file_path, style):
+            if self.git_add:
+                git_add(file_path)
+                # TODO: when git fails, it shows error and the info that
```
removed, pmt modify doesn't support this flag anymore

MartinBasti commented on tests/actions/modify/test_task_cli.py:
```diff
@@ -0,0 +1,741 @@
+
+
+@pytest.fixture
+def component_pipeline_dir():
```
I'm not going to update it to component_a and b repo fixtures, I have different use cases and expect to change it even more in future, with more types of samples. I don't think that a global example will work.

MartinBasti commented on src/pipeline_migration/yamleditor.py:
```diff
@@ -0,0 +1,532 @@
+
+        last_node, _ = path_stack[-1]
+
+        assert isinstance(last_node, (dict, list)) and hasattr(last_node, "lc")
```
linters require hasattr, or I'd need to rework everything into CommnetedSeq or CommentedMap

MartinBasti commented on src/pipeline_migration/yamleditor.py:
```diff
@@ -0,0 +1,532 @@
+            self._data = load_yaml(self.yaml_file_path, self.style)
+        return self._data
+
+    @data.deleter
+    def data(self):
+        self._data = None
+
+    def invalidate_yaml_data(self):
+        """Invalidate loaded yaml data.
+        After each change to yaml file, data must be invalidated"""
+        del self.data
```
https://github.com/konflux-ci/pipeline-migration-tool/pull/41#discussion_r2288395653

I already mentioned there, that I disagree with the change, as it's a property thus it has defined deleter. Also why it should be renamed to refresh? Function doesn't do refresh, it just invalidates data

MartinBasti commented on src/pipeline_migration/yamleditor.py:
```diff
@@ -0,0 +1,532 @@
+
+        Returns new path_stack to be used and data; If no change is needed,
+        original path_stack and data are returned
+        """
```
fixed

brunoapimentel commented on src/pipeline_migration/yamleditor.py:
```diff
@@ -105,6 +105,16 @@ def insert(self, path: YAMLPath, data: dict | list):
         last_node, _ = path_stack[-1]
 
+        if is_flow_style_seq(last_node):
+            # ve must update the parent via replacing
```
```suggestion
            # we must update the parent via replacing
``` 
(one more instance of this below)

MartinBasti replied:
fixed
